compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -1118,53 +1123,54 @@ private STNode parseFuncDefOrFuncTypeDesc(STNode metadata, STNode visibilityQual
-                        name, resourcePath, funcSignature, isObjectMember, isObjectTypeDesc);
-                endContext();
-                return funcDef;
+                return parseFunctionDefinition(metadata, visibilityQualifier, resourcePath, qualifiers, functionKeyword,
+                                               isObjectMember, isObjectTypeDesc, name);
             case DOT_TOKEN:
             case IDENTIFIER_TOKEN:
             case OPEN_BRACKET_TOKEN:
-                name = parseFunctionName();
                 resourcePath = parseRelativeResourcePath();
                 if (!isObjectMember) {
                     addInvalidNodeToNextToken(resourcePath,
                                               DiagnosticErrorCode.ERROR_RESOURCE_PATH_IN_FUNCTION_DEFINITION);
                     resourcePath = STNodeFactory.createEmptyNodeList();
                 }
-                switchContext(ParserRuleContext.FUNC_DEF);
-                funcSignature = parseFuncSignature(false);
-                funcDef = parseFuncDefOrMethodDeclEnd(metadata, visibilityQualifier, qualifiers, functionKeyword,
-                        name, resourcePath, funcSignature, isObjectMember, isObjectTypeDesc);
-                endContext();
-                return funcDef;
+                return parseFunctionDefinition(metadata, visibilityQualifier, resourcePath, qualifiers, functionKeyword,
+                                               isObjectMember, isObjectTypeDesc, name);
             case EQUAL_TOKEN:
             case SEMICOLON_TOKEN:
                 endContext();
-                return parseVarDeclWithFunctionType(metadata, visibilityQualifier, qualifiers, functionKeyword,
-                                                    STNodeFactory.createEmptyNode(), isObjectMember, isObjectTypeDesc,
-                                                    false);
+                functionKeyword = validateVisibilityQualifiersOfFunctionType(visibilityQualifier, qualifiers,
+                                                                             functionKeyword, isObjectMember);
+                List<STNode> extractQualifiersList = extractQualifiersOfFunctionType(qualifiers, isObjectMember,
+                                                                                     isObjectTypeDesc);
+                STNode qualifierNodeList = createFuncTypeQualNodeList(qualifiers, false);
+                STNode typeDesc = STNodeFactory.createFunctionTypeDescriptorNode(qualifierNodeList, functionKeyword,
+                                                                                 STNodeFactory.createEmptyNode());
+                // Check if it is a complex type desc starting with function type.
+                typeDesc = parseComplexTypeDescriptor(typeDesc, ParserRuleContext.TOP_LEVEL_FUNC_DEF_OR_FUNC_TYPE_DESC,
+                                                      false);
```

Comment:
We don't need this here right? since we know its not a complex type desc

Reply:
Fixed.